### PR TITLE
Added support for better pending pods check before failing them

### PR DESCRIPTION
**Pull request by agopala95**
Made changes within PollClusterHealth to only reflect on pending pod failures if the same pod remains in 'pending' state 10 times in a row. Otherwise, pending pod alerts are ignored in the clean run iteration. Key points to note here:-
1. The pending pod threshold for the same pod failing 10 times is set by the constant "pendingPodThreshold" in pkg/common/cluster/clusterutil.go.
2. In pkg/common/cluster/healthchecks/pods.go, the function checkPods returns the list of pending pods which is used by the function CheckClusterPodHealth which then returns the list of pending pods if they exist rather than returning the bool of the length of this list. This list is used by PollClusterHealth to keep track of the pending pods to see if any pod crosses the given threshold of being in pending state X times by calling the function CheckPendingPods. The map variable podErrorCount is used to keep track of a pod and its counter.
3. The pending pod check support is currently only fixed for the default e2e test suite. Support for e2e workload suites - redmine, guestbook and the managed upgrade workload can adopt the same strategy if needed.
4. The function CheckPodHealth is used by the above workload suites. As it uses the same checkPods function which returns a pod list rather than a bool of the length of this list, I've retained the bool in this function by checking the length. In the future, this can also be changed if support is needed for the workload suites.
5. The function CheckPendingPods is used by PollClusterHealth to keep track of pending pods and their corresponding counters of how many times they have remained under "pending state" consecutively. It takes in the pod list returned in each inner loop of PollClusterHealth and goes through each pod to see if it crosses the threshold (pendingPodThreshold) parameter. The podErrorCount variable which is also passed as a parameter is modified in each call to reflect the current state. This variable keeps track of pods and their current pending state counters.

### CODE REVIEW

**Comment by whereswaldon** [comment]
/retest

The problem was something I temporarily broke in the build system

File: pkg/common/cluster/clusterutil.go [hunk 1]
```diff
@@ -29,9 +29,12 @@ import (
 
 const (
 	// errorWindow is the number of checks made to determine if a cluster has truly failed.
-	errorWindow = 20
+	errorWindow         = 20
+	pendingPodThreshold = 10
```
Comment (whereswaldon) [inline pkg/common/cluster/clusterutil.go:33]: Can you document this constant?

**Review by whereswaldon — CHANGES_REQUESTED** [review]
I have a few documentation tweaks that I'd like, as well as some higher-level questions.

File: pkg/common/cluster/healthchecks/pods_test.go [hunk 2]
```diff
@@ -32,29 +32,25 @@ func TestCheckPodHealth(t *testing.T) {
...
-		{"one pod good one pod bad diff namespace", false, true, []runtime.Object{pod("a", ns1, v1.PodFailed), pod("b", ns2, v1.PodRunning)}},
-		{"single pod running", true, false, []runtime.Object{pod("a", ns1, v1.PodRunning)}},
-		{"single pod succeeded", true, false, []runtime.Object{pod("a", ns1, v1.PodSucceeded)}},
-		{"two succeeded pods diff namespace", true, false, []runtime.Object{pod("a", ns1, v1.PodSucceeded), pod("b", ns2, v1.PodSucceeded)}},
-		{"two running pods diff namespace", true, false, []runtime.Object{pod("a", ns1, v1.PodRunning), pod("b", ns2, v1.PodRunning)}},
+		{"no pods", 0, true, nil},
+		{"single pod failed", 0, true, []runtime.Object{pod("a", ns1, v1.PodFailed)}},
+		{"pod is pending beyond specified threshold", 1, false, []runtime.Object{pod("a", ns1, v1.PodPending)}},
+		{"single pod running", 0, false, []runtime.Object{pod("a", ns1, v1.PodRunning)}},
+		{"single pod succeeded", 0, false, []runtime.Object{pod("a", ns1, v1.PodSucceeded)}},
```
[the remark this answers is not in the document]
Comment (agopala95) [reply pkg/common/cluster/healthchecks/pods_test.go:44]: Yup, on second thought I guess the dropped tests are still valid as these do not seem to be impacted by the incoming pending pods check. I'll put them right back in.

File: pkg/common/cluster/clusterutil.go (hunk 1 above)
Comment (agopala95) [reply pkg/common/cluster/clusterutil.go:33]: I've updated the PR with a description on this.

File: pkg/common/cluster/healthchecks/pods.go [hunk 3]
```diff
@@ -14,18 +14,19 @@ import (
 
 // CheckClusterPodHealth attempts to look at the state of all internal cluster pods and
 // returns true if things are healthy.
```
[the remark this answers is not in the document]
Comment (agopala95) [reply pkg/common/cluster/healthchecks/pods.go:16]: I've updated the PR with an brief explanation on this.

File: pkg/common/cluster/healthchecks/pods.go [hunk 4]
```diff
@@ -98,3 +99,19 @@ func filterPods(podList *kubev1.PodList, predicates ...PodPredicate) *kubev1.Pod
 	}
 	return filteredPods
 }
+
+// CheckPendingPods looks for a pod that still remains under pending state for a given number of times in a row
+func CheckPendingPods(podlist []kubev1.Pod, podErrorCount map[string]int, pendingPodThreshold int) (map[string]int, bool, error) {
```
[the remark this answers is not in the document]
Comment (agopala95) [reply pkg/common/cluster/healthchecks/pods.go]: Yup, you're right. I've updated the code to reflect this change. I've removed returning the map from this function.

File: pkg/common/cluster/healthchecks/pods.go [hunk 5]
```diff
@@ -37,40 +38,40 @@ func CheckPodHealth(podClient v1.CoreV1Interface, logger *log.Logger, ns string,
...
-	foundErrorPods, err := checkPods(podClient, logger, filters...)
+	podlist, err := checkPods(podClient, logger, filters...)
 	if err != nil {
 		return false, err
 	}
-	return !foundErrorPods, err
+	return !(len(podlist) > 0), err
 }
 
 // checkPods looks for pods matching the supplied predicates and returns true if any are found
```
[the remark this answers is not in the document]
Comment (agopala95) [reply pkg/common/cluster/healthchecks/pods.go:48]: Yup. You can see note number 4 in the PR description.

File: pkg/common/cluster/healthchecks/pods.go [hunk 6]
```diff
@@ -98,3 +99,19 @@ func filterPods(podList *kubev1.PodList, predicates ...PodPredicate) *kubev1.Pod
 	}
 	return filteredPods
 }
+
+// CheckPendingPods looks for a pod that still remains under pending state for a given number of times in a row
```
Comment (whereswaldon) [inline pkg/common/cluster/healthchecks/pods.go:103]: I think this needs to say that you should re-use the podErrorCount parameter in subsequent calls. That's not intuitively obvious (at least to me). One way to reorganize the code to make this clearer would be to make this a method on a stateful type:

```go
type PodErrorTracker struct {
    Counts map[string]int
    MaxPendingPodsThreshold int
}
// CheckPendingPods checks each pod in the provided list for pending state and updates
// the PodErrorTracker accordingly. It returns nil if no pods were pending more than
// their maximum threshold, and errors if a pod ever exceeds its maximum
// pending threshold.
func (p *PodErrorTracker) CheckPendingPods(podList []kubev1.Pod) (error) {
// implementation details here
}
```

**Review by whereswaldon — CHANGES_REQUESTED** [review]
Thanks for the doc updates and for restoring those test cases! I have one more thing I'd like to look at before we call this done.

**Review by whereswaldon — APPROVED** [review]
/lgtm

**Review by 2uasimojo — COMMENTED** [review]
Sorry I missed this PR in flight. Having looked it over now, I have a couple of observations/concerns:

---

The idea here is to fail if the _same_ pod stays Pending 10x _in a row_, right?

Since
- The `PodErrorTracker` is a single instance per outer loop -- i.e. counts therein never get reset.
- The `Counts` map is indexed by pod name.

...this means we're relying on
- A Pod that leaves the Pending state will never return to the Pending state (I _think_ this is legit?)
- No two pods will ever have the same name.

I think this latter assumption may be faulty, though: for pods whose names aren't generated/templated.

I can't think of a real example off the top of my head, but theoretically, let's say there's some operator that periodically creates a Pod named `foo` to do some work. So we could see separate instances of `foo` in Pending state at various times during the health check loop and count them all as the same Pod.

An easy fix would be to index with the Pod's UUID instead of its name.

---

Another concern is that, if a Pending Pod appears on the last (or ninth-to-last) iteration of the outer loop, we'll still declare healthy at the end, even though there's a possibility that Pod will never complete.

I think the solution here would be to
1. clear counts for any Pods we don't increment in a given iteration; and
2. refuse to declare healthy until we hit an iteration where the map is empty.

(This could result in the outer loop needing to run up to 29x, which is fine.)

**Comment by whereswaldon** [comment]
@2uasimojo Good catch! I agree that the pod UUID would be a much better map key for tracking the counts.

As for a pending pod appearing towards the end, you're right again. Well spotted, and your solution sounds reasonable!

Thanks for looking over this in such depth! Sorry I missed these.

**Pull request merged**